flocker/control/test/test_persistence.py

```diff
@@ -249,3 +338,199 @@ def test_complex_keys(self):
+        if config['version'] != 2:
+            raise ConfigurationMigrationError(
+                "Supplied configuration was not a valid v2 config."
+            )
+        return json.dumps({"version": 3, "configuration": "fake"})
+
+
+class MigrateConfigurationTests(SynchronousTestCase):
+    """
+    Tests for ``migrate_configuration``.
+    """
+    v1_config = json.dumps({"version": 1})
+
+    def test_error_on_undefined_migration_path(self):
+        """
+        A ``MissingMigrationError`` is raised if a migration path
+        from one version to another cannot be found in the supplied
+        migration class.
+        """
+        e = self.assertRaises(
+            MissingMigrationError,
+            migrate_configuration, 1, 4, self.v1_config, StubMigration
+        )
+        expected_error = (
+            u'Unable to find a migration path for a version 3 to '
+            u'version 4 configuration. No migration method '
+            u'upgrade_from_v3 could be found.'
+        )
+        self.assertEqual(e.message, expected_error)
+
+    def test_sequential_migrations(self):
+        """
+        A migration from one configuration version to another will
+        sequentially perform all necessary upgrades, e.g. v1 to v2 followed
+        by v2 to v3.
+        """
+        # Get a valid v2 config.
+        v2_config = migrate_configuration(1, 2, self.v1_config, StubMigration)
+        # Perform two sequential migrations to get from v1 to v3, starting
+        # with a v1 config.
+        result = migrate_configuration(1, 3, self.v1_config, StubMigration)
+        # Compare the v1 --> v3 upgrade to the direct result of the
+        # v2 --> v3 upgrade on the v2 config, Both should be identical
+        # and valid v3 configs.
+        self.assertEqual(result, StubMigration.upgrade_from_v2(v2_config))
+
+
+UUIDS = st.basic(generate=lambda r, _: UUID(int=r.getrandbits(128)))
+
+DATASETS = st.builds(Dataset, dataset_id=UUIDS, maximum_size=st.integers())
+
+# Constrain primary to be True so that we don't get invariant errors from Node
+# due to having two differing manifestations of the same dataset id.
+MANIFESTATIONS = st.builds(
+    Manifestation, primary=st.just(True), dataset=DATASETS)
+IMAGES = st.builds(DockerImage, tag=st.text(), repository=st.text())
+NONE_OR_INT = st.one_of(
+    st.none(),
+    st.integers()
+)
+ST_PORTS = st.integers(min_value=1, max_value=65535)
+PORTS = st.builds(
+    Port,
+    internal_port=ST_PORTS,
+    external_port=ST_PORTS
+)
+LINKS = st.builds(
+    Link,
+    local_port=ST_PORTS,
+    remote_port=ST_PORTS,
+    alias=st.text(alphabet=string.letters, min_size=1)
+)
+FILEPATHS = st.text(alphabet=string.printable).map(FilePath)
+VOLUMES = st.builds(
+    AttachedVolume, manifestation=MANIFESTATIONS, mountpoint=FILEPATHS)
+APPLICATIONS = st.builds(
+    Application, name=st.text(), image=IMAGES,
+    # A MemoryError will likely occur without the max_size limits on
```

Comment:
Ah, interesting. I don't have any insights off the top of my head, and it's been a while since I've debugged Python memory usage errors. (https://pypi.python.org/pypi/memory_profiler seems a good place to start though). 

It might be interesting to try to make a minimal example that reproduces the memory crash, and submit it as a bug against Hypothesis. But, not necessary, and probably not a good use of ClusterHQ time.